### ISSUE TITLE
Spurious syntax error for certain switch-case-if scenario

### ISSUE THREAD

**Issue by lucaswerkmeister** [microsoft/TypeScript#7404]
**TypeScript Version:** 

master (a5354608fcafd986dcede6aa9832431a23cd3cdf), built from source. Also occurs on current [Playground](http://www.typescriptlang.org/Playground#src=switch%20%281%29%20{%0Acase%20%28a%29%3A%20{%0A%20%20%20%20if%20%28c%29%20{}%0A}%0A}%0A).

**Code**

``` ts
switch (1) {
case (a): {
    if (c) {}
}
}
```

**Expected behavior:**

Error about the unknown names `a` and `c`. Compiles to exactly the same JS, since the code is already plain JavaScript.

**Actual behavior:** 

Syntax errors.

> tmp.ts(3,12): error TS1005: ';' expected.
> tmp.ts(4,1): error TS1005: ':' expected.
> tmp.ts(5,1): error TS1128: Declaration or statement expected.

On Playground, you get some more errors (looks like CLI aborts at the syntax errors while Playground continues with type checking).

> <div>Type '(a: any) =&gt; { if(c: any): any; }' is not assignable to type 'number'.</div>
> 
> <div>A function whose declared type is neither 'void' nor 'any' must return a value.</div>

Also, `a` is documented as “parameter”. So apparently the parser thinks this is a function expression, not a case clause and following block.

This is fairly fragile. Many changes to the code will result in the error vanishing, including:
- Changing `a` to `1`
- Changing `c` to `1`
- Removing the parentheses around `a`
- Adding an extra pair of parentheses around `a`
- Removing the curly braces of the `case` block
- Replacing the empty `if` block with the empty statement `;`
- Replacing the empty `if` block with the nonempty statement `console.log("foo");`
- Completely removing the `if` statement

**Comment by DanielRosenwasser** [comment]
What's happening here is that you have what appears to be the beginning of an arrow function with a type annotation. `(x):`. Unfortunately, that block body looks like the beginning of a type literal, so we erroneously parse that out as a type, assuming that you just have syntactic errors in your block (see #2496 for a similar issue).

Unfortunately this ends up breaking valid code. Any ideas @vladima or @CyrusNajmabadi?

**Comment by CyrusNajmabadi** [comment]
> Any ideas ... ?

Here's what the EcmaScript grammar has to say about this production:

```case Expression[In, ?Yield] : StatementList[?Yield, ?Return]opt```

So technically, you could have a ```case <arrow function>:```  (though it would be pretty nonsensical).  I can think of a few ways to do things:

1. Do speculative lookahead to decide what sort of expressoin this is, and parse the code appropriately.  Pros: correct.  Cons: slower.
2. Tighten up the existing speculative lookahead for arrow functions to not match here since there is no ```=>```.  Pros: improves this situation.  Cons: may miss some other.  Also, the lookahead on arrow functions is already pretty complex.
3. Don't allow parsing arrow functions here.  Pros: nice and simple and pretty much always what you want.  Cons: not matching emascript spec anymore.

I'd probably just go with '2'.  It's a nice and narrow fix that gets the result you want here with a minimum of fuss.

**Comment by CyrusNajmabadi** [comment]
@DanielRosenwasser Didn't you have to deal with the object-literal/object-tyep ambiguity before?  Did you write any sort of helper to help out here?

**Comment by vladima** [comment]
there are other cases when we prefer parse-and-recover case for error function to legal scenarios.
Now this is an error:
```ts
(a / b
)
{
    
}
```
but this is not:
```ts
(a / b)
{
    
}

```

**Comment by CyrusNajmabadi** [comment]
hrmm:   Why is:  

```
(a / b
)
{

}
```

An error?  I'm not quite sure i got the reasoning there.  Thanks!

**Comment by vladima** [comment]
first we try to parse expression as parenthesized arrow function:
- `(` - ok, proceed with parameter list
- parse parameter name, got `a` - ok
- try parse type annotation, current token is not `:` - do nothing
- try parse initializer, current token is not `=`, assume that user missed this token, show error that `=` is expected and continue
- parse expression in initializer - current token is `/` - rest of the line is parsed as unterminated regex literal
- current token is `)` - done with a parameter list
- current token is `{` - assume that used has missed `=>` and continue.

As a result we have parsed 
```ts
(a / b
)
```
as a beginning of parenthesized arrow function

**Comment by RyanCavanaugh** [comment]
Similar to weird constructions that can end up looking like type arguments, this is an edge case deviation from JS semantics that I think we just have to bite the bullet on. This program
```ts
let string = 0, a = 1;
switch (1) {
  case (a): string => null;
}
```
In JS parses as
```ts
let string = 0, a = 1;
switch (1) {
  case a:
    string => null;
}
```
but TypeScript sees it as
```ts
let string = 0, a = 1;
switch (1) {
  case ((a): string => null)
}
```
and expects a `:`. Without an impractically large amount of lookahead I don't think this is something we can fix.

**Comment by RyanCavanaugh** [comment]
@DanielRosenwasser pointed out that we just have to disallow parens following `case` as being interpeted as parameter lists, which isn't too weird of a change and wouldn't stop any plausibly-correct code from parsing. Still moving this to Dormant but someone can send a PR if they like
